thread/thread11.h

```diff
@@ -24,160 +25,49 @@ limitations under the License.
-        {
-            parent = CURRENT;
-//            LOG_DEBUG("arguments stored in tuple");
-        }
-
-        static return_type stub11(void* args)
-        {
-            typedef ThreadContext11__ Context;
-            auto ctx_ = (Context*)args;
-            Context ctx = std::move(*ctx_);
-//            LOG_DEBUG("arguments tuple moved");
-            ctx_->got_it = true;
-            thread_yield_to(ctx.parent);
-            return tuple_assistance::apply(ctx.start, ctx.args);
-        }
-        thread* thread_create(thread_entry start_)
-        {
-            auto th = ::photon::thread_create(start_, this, stack_size);
-            thread_yield_to(th);
-            return th;
-        }
-    };
+namespace photon {
 
-    template<typename F>
-    struct ThreadContext11 : public ThreadContext11__<F>
-    {
-        typedef ThreadContext11__<F> base;
-        using base::base;
-        using base::stub11;
-        static void* stub(void* args)
-        {
-            stub11(args);
-            return nullptr;
-        }
-        thread* thread_create()
-        {
-            return base::thread_create(&stub);
-        }
-    };
-
-    template<typename...PARAMETERS>
-    struct ThreadContext11<void* (*)(PARAMETERS...)> :
-        public ThreadContext11__<void* (*)(PARAMETERS...)>
-    {
-        typedef void* (*F)(PARAMETERS...);
-        typedef ThreadContext11__<F> base;
-        using base::base;
-        using base::stub11;
-        static void* stub(void* args)
-        {
-            return stub11(args);
-        }
-        thread* thread_create()
-        {
-            return base::thread_create(&stub);
-        }
-    };
-
-    template <typename F, typename... ARGUMENTS>
-    inline std::enable_if_t<is_function_pointer<F>::value, thread*>
-    thread_create11(uint64_t stack_size, F f, ARGUMENTS&&... args) {
-        return ThreadContext11<F>(stack_size, f,
-                                  std::forward<ARGUMENTS>(args)...)
-            .thread_create();
-    }
-
-    template <typename F, typename... ARGUMENTS>
-    inline std::enable_if_t<is_function_pointer<F>::value, thread*>
-    thread_create11(F f, ARGUMENTS&&... args) {
-        return thread_create11<F, ARGUMENTS...>(
-            DEFAULT_STACK_SIZE, f, std::forward<ARGUMENTS>(args)...);
-    }
-
-    template <typename CLASS, typename F, typename... ARGUMENTS>
-    inline std::enable_if_t<std::is_member_function_pointer<F>::value, thread*>
-    thread_create11(uint64_t stack_size, F f, CLASS* obj, ARGUMENTS&&... args) {
-        auto pmf = ::get_member_function_address(obj, f);
-        return thread_create11(stack_size, pmf.f, pmf.obj,
-                               std::forward<ARGUMENTS>(args)...);
-    }
+template<class F, class... Args, size_t I0, size_t... I>
+void __thread_execute(std::tuple<F, Args...>& t, std::index_sequence<I0, I...>) {
+    std::__invoke(std::move(std::get<I0>(t)), std::move(std::get<I>(t))...);
```

Comment:
`std::__invoke` is a kind of STL internal component, these underscores indicate that it should not supposed to be used in non-STL implemention.

Reply:
> `std::__invoke` is a kind of STL internal component, these underscores indicate that it should not supposed to be used in non-STL implemention.

We can implement the invoke function ourselves